### PR TITLE
DROOLS-2652: [DMN Designer] Different icons for different connector types

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-2652

This PR is incomplete at the moment (tests need writing, and I need the final SVGs from UX). It is for review by @romartin.... @romartin This PR contains the changes we were talking about for different glyphs for different purposes. I'll call out the changes in Stunner so you can review better. 

I wanted to check you were OK with the proposal sooner than later.

### CODE REVIEW

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/palette/BS3PaletteWidgetImpl.java [hunk 1]
```diff
@@ -238,7 +238,14 @@ public void onDragProxyComplete(final String definitionId,
...
     @Override
     public Glyph getShapeGlyph(final String definitionId) {
-        return getShapeFactory().getGlyph(definitionId);
+        return getShapeFactory().getGlyph(definitionId,
+                                          AbstractPalette.PaletteGlyphConsumer.class);
+    }
+
+    @Override
+    public Glyph getShapeDragProxyGlyph(final String definitionId) {
+        return getShapeFactory().getGlyph(definitionId,
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/palette/BS3PaletteWidgetImpl.java]: @romartin I added the ability to have a different glyph for the palette item and drag proxy; if needed.

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/shape/factory/ShapeFactory.java [hunk 2]
```diff
@@ -50,4 +50,24 @@
      * @return
      */
     Glyph getGlyph(String definitionId);
+
+    /**
+     * Returns the glyph (thumbnail/miniature) for the specified Definition identifier and consumer.
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/shape/factory/ShapeFactory.java]: @romartin This is the centre of the changes.. the ability to get a glyph (from a ```ShapeFactory```) for different "consumers" (of the glyph). The default implementation ignores the consumer.

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/definition/shape/ShapeDef.java [hunk 3]
```diff
@@ -36,4 +38,11 @@ default Glyph getGlyph(final Class<? extends W> type,
                            final String defId) {
         return ShapeGlyph.create();
     }
+
+    @SuppressWarnings("unused")
+    default Glyph getGlyph(final Class<? extends W> type,
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/definition/shape/ShapeDef.java]: @romartin ```DelegateShapeFactory``` gets glyphs from the ```ShapeDef``` so provide similar support.

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/shape/factory/ShapeFactory.java (hunk 2 above)
Comment (romartin) [reply kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/shape/factory/ShapeFactory.java]: Ahh ok I see so the consumer is just a class, a type, which acts like a marker. I mean there is no expected logic in that type, it's just for the shape factory instance in order to provide different glyphs.. right?
In that case, there exist "common" consumers? Or for now it's just DMN who uses it?

**Comment by romartin** [comment]
Hey @manstis 
Sorry for the late review. It sounds good for me yeah, not much to say... thanks! :) 
PS Anyway let us know when the PR is ready for review, and will do more code review and  some tests also. Thanks!

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/shape/factory/ShapeFactory.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/shape/factory/ShapeFactory.java]: Yes just a marker; I thought more extensible than an enum.

Yes there are defaults in Stunner for Palette, Palette drag proxy and Action Toolbox. The default (old, original method) is still used for Canvas.

**Comment by jomarko** [comment]
@manstis This is still not ready for review?

**Comment by manstis** [comment]
@jomarko I'll check test coverage today; but I'm still waiting for Brian Dellascio from UX to provide the icons before this can be merged :-(

**Comment by jomarko** [comment]
@manstis ok, understood, no problem.

**Comment by manstis** [comment]
@jomarko @romartin This PR is "complete" other than waiting for some PNGs from UX.

Feel free to review the code/tests; or wait until I advise the PNGs are committed.. up to you :-)

**Comment by manstis** [comment]
@jomarko @romartin This PR is "good to go" now (new images received from UX).

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/resources/DMNSVGGlyphFactory.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    ImageDataUriGlyph BUSINESS_KNOWLEDGE_MODEL_PALETTE = ImageDataUriGlyph.create(DMNImageResources.INSTANCE.businessKnowledgeModelGlyph().getSafeUri());
+
+    ImageDataUriGlyph DECISION_PALETTE = ImageDataUriGlyph.create(DMNImageResources.INSTANCE.decisionGlyph().getSafeUri());
+
+    ImageDataUriGlyph INPUT_DATA_PALETTE = ImageDataUriGlyph.create(DMNImageResources.INSTANCE.inputDataGlyph().getSafeUri());
+
+    ImageDataUriGlyph KNOWLEDGE_SOURCE_PALETTE = ImageDataUriGlyph.create(DMNImageResources.INSTANCE.knowledgeSourceGlyph().getSafeUri());
+
+    ImageDataUriGlyph TEXT_ANNOTATION_PALETTE = ImageDataUriGlyph.create(DMNImageResources.INSTANCE.textAnnotationGlyph().getSafeUri());
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/resources/DMNSVGGlyphFactory.java]: Just question, is it intentional that constants wit the suffix `PALETTE` call methods with suffix `Glyph()` and constants with suffix `GLYPH` call methods with suffix `Toolbox()`? It is probably correct just asking to be sure.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/shape/def/DMNSVGShapeDefImplTest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        assertEquals(DMNSVGGlyphFactory.KNOWLEDGE_SOURCE_GLYPH,
+                     shapeDef.getGlyph(KnowledgeSource.class, DEFINITION_ID));
+        assertEquals(DMNSVGGlyphFactory.TEXT_ANNOTATION_GLYPH,
+                     shapeDef.getGlyph(TextAnnotation.class, DEFINITION_ID));
+
+        assertTrue(shapeDef.getGlyph(DMNViewDefinition.class, DEFINITION_ID) instanceof ShapeGlyph);
+    }
+
+    @Test
+    public void testGetGlyphWithConsumer() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/shape/def/DMNSVGShapeDefImplTest.java]: Should we also in this test check unknown type? I mean something similar we did on last line in the test above.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/shape/def/DMNSVGShapeDefImplTest.java]: Added in new commit.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/resources/DMNSVGGlyphFactory.java (hunk 4 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/resources/DMNSVGGlyphFactory.java]: Updated to be more consistent throughout.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by jomarko** [comment]
@manstis I saw your conversation with UX team, is the thing with greyed icons (when they are not hovered) solved? Or will be done as separate task?

**Comment by manstis** [comment]
@jomarko It has not been solved and remains a generic Stunner issue (or requirement to change to UX's needs). I have proposed any outstanding UX issues with the Toolbox and related Tooltips be tracked by https://issues.jboss.org/browse/JBPM-7153

**Pull request merged**